code/modules/reagents/cocktails.dm

```diff
@@ -0,0 +1,477 @@
+// This is a system used to change the name and description of a glass of
+// alcohol if it meets certain minimum proportions of ingredients. It 
+// replaces the previous system, which used chemical reactions.
+
+/decl/cocktail
+	var/name                   // Cocktail name, applied to the glass.
+	var/description            // Cocktail list, applied to the glass.
+	var/list/ratios            // Associative list of reagents to a minimum percentage of mix (<1).
+	                           // Reagents with no assoc value will count as valid for any amount (even 0.001u)
+	var/order_specific = FALSE // If set, cocktail will fail if ingredients were added out of ratio order.
+
+// Shoot for total ratios of about 70% (0.7) for any cocktail that doesn't need 
+// to be super precise - this will leave room in a mix for people to spike your 
+// drink or to be comfortably over or under their proportions without having to 
+// be frustratingly picky with measurements.
+
+/decl/cocktail/proc/get_presentation_name(var/obj/item/prop)
+	. = name
+	if(prop?.reagents?.has_reagent(/decl/reagent/drink/ice) && !(/decl/reagent/drink/ice in ratios))
+		. = "[name], on the rocks"
+
+/decl/cocktail/proc/mix_priority()
+	. = length(ratios)
+
+/decl/cocktail/proc/matches(var/obj/item/prop)
+	if(length(ratios) > length(prop.reagents.reagent_volumes))
+		return FALSE
+	var/list/check_ratios
+	var/i = 0
+	for(var/rtype in ratios)
+		i++
+		if(!prop.reagents.has_reagent(rtype) || (order_specific && prop.reagents.reagent_volumes[i] != rtype))
+			return FALSE
+		if(isnum(ratios[rtype]))
+			LAZYSET(check_ratios, rtype, ratios[rtype])
+	var/effective_volume = prop.reagents.total_volume - REAGENT_VOLUME(prop.reagents, /decl/reagent/drink/ice)
```

Comment:
wot about cocktails where ice is an ingridient

Reply:
There is exactly one of them now. Can duplicate the check for it though.